File: 프로그래머스/LV2/코딩테스트 연습/연습문제/디펜스 게임/디펜스 게임.py
from heapq import heappush, heappop

def solution(n, k, enemy):
    heap = []
    sum = 0
    
    for i in range(len(enemy)):
        heappush(heap, enemy[i])

        if len(heap) > k:
            sum += heap[0]
            heappop(heap)

        if sum > n:
            return i

    return len(enemy)

File: 프로그래머스/LV2/코딩테스트 연습/연습문제/디펜스 게임/디펜스 게임.cpp
#include <string>
#include <vector>
#include <queue>
using namespace std;

int solution(int n, int k, vector<int> enemy) {
    int sum = 0;
    priority_queue<int, vector<int>, greater<int>> min_heap;

    for(int i = 0; i < enemy.size(); i++) {
        min_heap.push(enemy[i]);

        if(min_heap.size() > k) {
            sum += min_heap.top();
            min_heap.pop();
        }

        if(sum > n)
            return i;
    }

    return enemy.size();
}